client/fingerprint/dynamic_host_volumes.go

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: BUSL-1.1
+
+package fingerprint
+
+import (
+	"context"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+	hvm "github.com/hashicorp/nomad/client/hostvolumemanager"
+	"github.com/hashicorp/nomad/helper"
+)
+
+func NewPluginsHostVolumeFingerprint(logger hclog.Logger) Fingerprint {
+	return &DynamicHostVolumePluginFingerprint{
+		logger: logger.Named("host_volume_plugins"),
+	}
+}
+
+var _ ReloadableFingerprint = &DynamicHostVolumePluginFingerprint{}
+
+type DynamicHostVolumePluginFingerprint struct {
+	logger hclog.Logger
+}
+
+func (h *DynamicHostVolumePluginFingerprint) Reload() {
+	// host volume plugins are re-detected on agent reload
+}
+
+func (h *DynamicHostVolumePluginFingerprint) Fingerprint(request *FingerprintRequest, response *FingerprintResponse) error {
+	// always add "mkdir" plugin
+	h.logger.Debug("detected plugin built-in", "plugin_id", "mkdir", "version", hvm.HostVolumePluginMkdirVersion)
+	defer response.AddAttribute("plugins.dhv.version.mkdir", hvm.HostVolumePluginMkdirVersion)
+	response.Detected = true
+
+	// this config value will be empty in -dev mode
+	pluginDir := request.Config.DynamicHostVolumePluginPath
+	if pluginDir == "" {
+		return nil
+	}
+
+	plugins, err := GetHostVolumePluginVersions(h.logger, pluginDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			h.logger.Debug("plugin dir does not exist", "dir", pluginDir)
+		} else {
+			h.logger.Warn("error finding plugins", "dir", pluginDir, "error", err)
+		}
+		return nil // don't halt agent start
+	}
+
+	// if this was a reload, wipe what was there before
+	for k := range request.Node.Attributes {
+		if strings.HasPrefix(k, "plugins.dhv.") {
+			response.RemoveAttribute(k)
+		}
+	}
+
+	// set the attribute(s)
+	for plugin, version := range plugins {
+		h.logger.Debug("detected plugin", "plugin_id", plugin, "version", version)
+		response.AddAttribute("plugins.dhv.version."+plugin, version)
+	}
+
+	return nil
+}
+
+func (h *DynamicHostVolumePluginFingerprint) Periodic() (bool, time.Duration) {
+	return false, 0 // not periodic;  db TODO(1.10.0): could be?
+}
+
+// GetHostVolumePluginVersions finds all the executable files on disk
+// that respond to a Version call (arg $1 = 'version' / env $OPERATION = 'version')
+// The return map's keys are plugin IDs, and the values are version strings.
+func GetHostVolumePluginVersions(log hclog.Logger, pluginDir string) (map[string]string, error) {
+	files, err := helper.FindExecutableFiles(pluginDir)
+	if err != nil {
+		return nil, err
+	}
+
+	plugins := make(map[string]string)
+	mut := sync.Mutex{}
+	var wg sync.WaitGroup
+
+	for file, fullPath := range files {
+		wg.Add(1)
+		go func(file, fullPath string) {
+			defer wg.Done()
+			// really should take way less than a second
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			defer cancel()
+
+			log := log.With("plugin_id", file) // db TODO: "name" like CNI, instead of "ID" ?
```

Comment:
What do you mean by this TODO? I think we probably want to force the plugin ID to be the file name, right?

Reply:
I was considering maybe doing this all over the place

```suggestion
			log := log.With("name", file)
```

similar to CNI networks and plugins, e.g.

```
[DEBUG] fingerprint/cni.go:76: client.fingerprint_mgr: detected CNI network: name=cool
[DEBUG] fingerprint/plugins_cni.go:119: client.fingerprint_mgr.cni_plugins: failed to parse CNI plugin version: name=cool
```

Reply:
Up to you. `plugin_id` matches the field in the volume spec though.

Reply:
yeah, I had basically talked myself out of it by writing my comment, and you helped tip it over the edge :yum:  I'll keep plugin_id.